File: functions/src/ai/transcribe.ts
import { HttpsError } from "firebase-functions/v2/https";
import { loadLlmKeys } from "./keys";

const MAX_AUDIO_BYTES = 4 * 1024 * 1024;
const OPENAI_STT_URL = "https://api.openai.com/v1/audio/transcriptions";
const XAI_STT_URL = "https://api.x.ai/v1/stt";

export interface TranscribeRequest {
  audioBase64?: string;
  mimeType?: string;
  filename?: string;
}

function safeFilename(raw: unknown, mimeType: string): string {
  if (typeof raw === "string") {
    const base = raw.split(/[/\\]/).pop()?.trim() ?? "";
    if (/^[\w.-]+\.(webm|wav|mp3|mp4|m4a|ogg|mpeg|mpga)$/i.test(base)) {
      return base;
    }
  }
  const lower = mimeType.toLowerCase();
  if (lower.includes("wav")) return "chunk.wav";
  if (lower.includes("mp4") || lower.includes("m4a")) return "chunk.m4a";
  if (lower.includes("ogg")) return "chunk.ogg";
  if (lower.includes("mpeg") || lower.includes("mp3")) return "chunk.mp3";
  return "chunk.webm";
}

function transcriptText(payload: unknown): string {
  if (!payload || typeof payload !== "object") return "";
  const text = (payload as { text?: unknown }).text;
  return typeof text === "string" ? text.trim() : "";
}

async function transcribeOpenAi(
  buffer: Buffer,
  mimeType: string,
  filename: string,
  apiKey: string,
): Promise<Response> {
  const form = new FormData();
  form.append("model", "whisper-1");
  form.append("file", new Blob([new Uint8Array(buffer)], { type: mimeType }), filename);
  return fetch(OPENAI_STT_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}` },
    body: form,
  });
}

async function transcribeXai(
  buffer: Buffer,
  mimeType: string,
  filename: string,
  apiKey: string,
  includeModel = true,
): Promise<Response> {
  const form = new FormData();
  if (includeModel) form.append("model", "grok-stt");
  form.append("language", "fr");
  form.append("format", "true");
  form.append("file", new Blob([new Uint8Array(buffer)], { type: mimeType }), filename);
  return fetch(XAI_STT_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}` },
    body: form,
  });
}

export async function runAiTranscribe(
  uid: string,
  data: TranscribeRequest,
): Promise<{ text: string }> {
  const keys = await loadLlmKeys(uid);
  if (!keys.openai && !keys.xai) {
    throw new HttpsError(
      "failed-precondition",
      "Transcription indisponible (service).",
    );
  }

  const audioBase64 = typeof data.audioBase64 === "string" ? data.audioBase64.trim() : "";
  if (!audioBase64) {
    throw new HttpsError("invalid-argument", "Audio manquant.");
  }

  let buffer: Buffer;
  try {
    buffer = Buffer.from(audioBase64, "base64");
  } catch {
    throw new HttpsError("invalid-argument", "Audio invalide.");
  }
  if (!buffer.length) {
    throw new HttpsError("invalid-argument", "Audio vide.");
  }
  if (buffer.length > MAX_AUDIO_BYTES) {
    throw new HttpsError("invalid-argument", "Audio trop volumineux.");
  }

  const mimeType =
    typeof data.mimeType === "string" && data.mimeType.trim()
      ? data.mimeType.trim()
      : "audio/webm";
  const filename = safeFilename(data.filename, mimeType);

  const attempts: Array<{
    name: "openai" | "xai";
    run: () => Promise<Response>;
  }> = [];
  if (keys.xai) {
    attempts.push({
      name: "xai",
      run: () => transcribeXai(buffer, mimeType, filename, keys.xai),
    });
  }
  if (keys.openai) {
    attempts.push({
      name: "openai",
      run: () => transcribeOpenAi(buffer, mimeType, filename, keys.openai),
    });
  }

  let lastKind: "service" | "network" = "network";
  for (const attempt of attempts) {
    let response: Response;
    try {
      response = await attempt.run();
    } catch {
      lastKind = "network";
      continue;
    }
    if (response.ok) {
      return { text: transcriptText(await response.json()) };
    }
    if (attempt.name === "xai" && response.status === 400) {
      try {
        const retry = await transcribeXai(buffer, mimeType, filename, keys.xai, false);
        if (retry.ok) {
          return { text: transcriptText(await retry.json()) };
        }
        response = retry;
      } catch {
        lastKind = "network";
        continue;
      }
    }
    lastKind = response.status === 401 || response.status === 403 ? "service" : "network";
  }

  throw new HttpsError(
    lastKind === "service" ? "failed-precondition" : "unavailable",
    lastKind === "service"
      ? "Transcription indisponible (service)."
      : "Transcription indisponible (réseau).",
  );
}

File: backend/app/api/transcribe.py
"""Live-notes STT — Whisper / Grok chunks for Electron (Web Speech is broken there)."""
from __future__ import annotations

import re

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile

from app.ai.stt import SttError, transcribe_audio
from app.core.auth_deps import require_firebase_user
from app.core.firebase import FirebaseUser

router = APIRouter(prefix="/api", tags=["transcribe"])

_MAX_AUDIO_BYTES = 4 * 1024 * 1024
_SAFE_NAME = re.compile(r"^[\w.-]+\.(webm|wav|mp3|mp4|m4a|ogg|mpeg|mpga)$", re.I)


def _filename(upload: UploadFile) -> str:
    name = (upload.filename or "").split("/")[-1].split("\\")[-1]
    if _SAFE_NAME.match(name):
        return name
    content = (upload.content_type or "").lower()
    if "wav" in content:
        return "chunk.wav"
    if "mp4" in content or "m4a" in content:
        return "chunk.m4a"
    if "ogg" in content:
        return "chunk.ogg"
    if "mpeg" in content or "mp3" in content:
        return "chunk.mp3"
    return "chunk.webm"


@router.post("/transcribe")
async def api_transcribe(
    file: UploadFile = File(...),
    user: FirebaseUser = Depends(require_firebase_user),
):
    data = await file.read()
    if not data:
        raise HTTPException(400, "Audio vide.")
    if len(data) > _MAX_AUDIO_BYTES:
        raise HTTPException(413, "Audio trop volumineux.")

    filename = _filename(file)
    mime = file.content_type or "application/octet-stream"

    try:
        text = await transcribe_audio(
            data,
            filename=filename,
            mime=mime,
            uid=user.uid,
        )
    except SttError as exc:
        raise HTTPException(503, str(exc)) from exc

    return {"text": text}
